### DISCUSSION TITLE
Fire events asynchronously

### DISCUSSION THREAD

**Discussion by J-Cake** [vizia/vizia#385] [Questions] [answered]
Heyy just wondering if/how it's possible to spawn events from outside the main thread?

Thanks

**Comment by geom3trik** [comment, marked as answer]
Hello,

There are a couple of ways to achieve this. The first is to use [`cx.spawn()`](https://docs.vizia.dev/vizia/context/struct.EventContext.html#method.spawn) which spawns a thread for you and provides a proxy context you can use to send events back to the main thread. For example:
```rust
cx.spawn(|cx|{
    // Do something here off the main thread
    cx.emit(...); // Send event back to main thread
})
```
The second method is if you prefer to spawn and manage threads yourself. You can use [`cx.get_proxy()`](https://docs.vizia.dev/vizia/context/struct.Context.html#method.get_proxy) to get a `ContextProxy` which you can move into a thread and then send events back to the main thread as above.

**Comment by J-Cake** [comment]
Aha yesss fantastic thank you

**Comment by uklotzde** [comment]
The ability to create a `ProxyContext` that implements `EmitContext` is also needed in `Model::event()` when you spawn an async task.

Unfortunately, `EventContext` does not provide the API for this. Instead, it is hidden in the opinionated `EventContext::spawn()` method.

But I don't want to spawn a system thread. I need to pass it into an `async move {}` block that is spawned on the async runtime. When finished this task is supposed to emit an event.

**Reply by geom3trik** [reply]
Sure, I can add this to `EventContext`

**Reply by uklotzde** [reply]
I expect that this is a common use case. An event triggers an asynchronous (side-)effect that submits its results by emitting another event when finished.

**Comment by uklotzde** [comment]
As a workaround I obtain the `ContextProxy` on startup and pass it around. Keeping this extra state would be avoidable.
